(post) Apache MOD test filtering functionality (#85)
puppetlabs/iac: _posts/2020-07-03-apache-mod-test-improvements.md

- [Fixes for one environment][pr-1913] having a knock-on effect for a [different environment][pr-2041] (thanks to [h-haaks][h-haaks] for the fix!)

When a new version of an OS comes out, it is quite common for issues to manifest when trying to configure some of the less mainstream [Apache MODs][apache-mod]:
- A simple change in it's package name (_thankfully a relatively quick fix!_)
- The package is no longer available from the core "out-of-the-box" repositories (e.g. a 3rd party repos or edge/experimental/extra package streams need to be configured)
- Package dependencies are unavailable in non-standard repositories
- Support is dropped altogether - there is no way to locate the package or dependencies for a newer OS version

This led us to a conversation about how we continue to support all the various prerequisites and configurations required when presented with a matrix of [Apache MODs][apache-mod] and [compatible OSs][compatible-platforms].
We could, if time permitted, ensure that we have the steps to install every version...of every MOD...on every OS...documented and implemented.

Time permitting...

## Solution
### Test Support Policy

> Straight away, we realised that, in future, we would need to strictly time box the amount of time we would spend attempting to configure the OS with the correct repositories and/or dependencies for a given MOD.

Reviewer: ```suggestion
Straight away, we realised that, in the future, we need to strictly time box the amount of time we spend attempting to configure the OS with the correct repositories and/or dependencies for any given MOD.
```

### We've agreed we're not supporting this MOD on this OS version - what now?
You may be familiar with [RSpec's conditional filtering][rspec-conditional-filtering]?
If not, [here is an example][rspec-filtering-example]:
```ruby
describe 'apache::mod::ldap', unless: os[:family] == 'redhat' && os[:release].to_i >= 8 do
```
This will mean that this test will **not** run if the OS family is `RedHat` and the version is `8` or higher.
These work well, but, things can start to get messy when there is quite a narrow band of compatible OSs for a MOD.
Consider [`v2.4.3` of `mod_auth_openidc`][mod_auth_openidc_2_4_3] - we can see from the build artifacts' names that we "officially" have support for:
- Ubuntu 16.04
- Ubuntu 18.04
- Debian 10
- RHEL 7

This does not consider the fact that there are many other OSs derived from RHEL and other distro vendors / maintainers can port packages to their OS.
For example, after a bit more digging, I can see that the `mod_auth_openidc` package is available for the following OSs too:
- CentOS 7
- CentOS 8 (via AppStream)
- RHEL 8 (via AppStream)
- Fedora 30
- Fedora 31
- Fedora 32

> This search was not exhaustive, and to make it so would require a lot more effort on my part, which is part of the issue we find ourselves up against.

Reviewer: ```suggestion
This search was not exhaustive, and to make it so would require a lot more effort, which is part of the issue we find ourselves up against.
```
